addons/mass_mailing/models/mail_notification.py

```diff
@@ -49,53 +40,59 @@ class MailingTrace(models.Model):
-        ("BOUNCE", "Email address rejected by destination"),
-        ("UNKNOWN", "Unknown error")], string='Failure type')
-    state_update = fields.Datetime(compute="_compute_state", string='State Update',
-                                   help='Last state update of the mail',
-                                   store=True)
-
-    @api.depends('trace_type', 'mass_mailing_id')
+    notification_status = fields.Selection(selection_add=[
+        ('outgoing', 'Outgoing'),
+        ('opened', 'Opened'),
+        ('replied', 'Replied')])
+    status_update = fields.Datetime(string='State Update', help='Last state update of the mail', store=True)
+
+    @api.depends('mass_mailing_id.campaign_id')
+    def _compute_campaign_id(self):
+        for notif in self:
+            notif.campaign_id = notif.mass_mailing_id.campaign_id
+
+
+    @api.depends('notification_type', 'mass_mailing_id', 'res_partner_id')
     def _compute_display_name(self):
         for trace in self:
-            trace.display_name = '%s: %s (%s)' % (trace.trace_type, trace.mass_mailing_id.name, trace.id)
-
-    @api.depends('sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled')
-    def _compute_state(self):
-        self.update({'state_update': fields.Datetime.now()})
-        for stat in self:
-            if stat.canceled:
-                stat.state = 'canceled'
-            elif stat.exception:
-                stat.state = 'exception'
-            elif stat.opened or stat.clicked:
-                stat.state = 'opened'
-            elif stat.replied:
-                stat.state = 'replied'
-            elif stat.bounced:
-                stat.state = 'bounced'
-            elif stat.sent:
-                stat.state = 'sent'
-            else:
-                stat.state = 'outgoing'
+            trace.display_name = '%s: %s (%s)' % (trace.notification_type,
+                                                  trace.mass_mailing_id.name or trace.res_partner_id.display_name or '',
+                                                  trace.id)
+
+    def _update_trace_status(self, values):
+        if set(['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled']) & set(values.keys()):
+            values['status_update'] = fields.Datetime.now()
+            if values.get('canceled', False):
+                values['notification_status'] = 'canceled'
+            elif values.get('exception', False):
+                values['notification_status'] = 'exception'
+            elif values.get('opened', False) or values.get('clicked', False):
+                values['notification_status'] = 'opened'
+            elif values.get('replied', False):
+                values['notification_status'] = 'replied'
+            elif values.get('bounced', False):
+                values['notification_status'] = 'bounced'
+            elif values.get('sent', False):
+                values['notification_status'] = 'sent'
```

Comment:
values.get('canceled') will return None if it doesn't contain 'canceled', and None is Falsy.

In other words, you can remove all the "False" default values.

Also, this can be refactored.

```
statuses_whitelist = ['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled']
if any(values.keys() in statuses_whitelist):
    notification_status = False
    if values.get('opened') or values.get('clicked'):
        notification_status = 'opened'
    else:
        notification_status = next(key for key in values.keys() if key in statuses_whitelist, False)
    if notification_status:
        values['notification_status'] = notification_status
```

And finally, the method should return a status rather than modifying one of its parameter.
Modifying method parameters is considered as bad practice.
I'll let you do the change for that though.

Reply:
Ok for the get False redundancy. 
But the refactor propose isn't correct, because the order has some importance.

Reply:
"next" will take the first occurrence.

So if you order your whitelist correctly, it should do the same.

Reply:
Yes, but the double condition `values.get('opened') or values.get('clicked')` is also important with the order. Maybe create a whitelist with tuple ([conditions] and 'value_to_set') and iterate it.

Reply:
Ok I see what you mean.

Then maybe just:

```
notification_status = next(key for key in values.keys() if key in statuses_whitelist, False)
if notification_status = 'clicked'
    notification_status = 'opened'
```

Since it seems to be a particular case for the "clicked".

Reply:
Yes good trick, nice. I am on it.